src/meta/src/barrier/schedule.rs

```diff
@@ -80,13 +79,27 @@ impl<S: MetaStore> BarrierScheduler<S> {
 
     /// Attach `new_notifiers` to the very first scheduled barrier. If there's no one scheduled, a
     /// default checkpoint barrier will be created.
-    async fn attach_notifiers(&self, new_notifiers: impl IntoIterator<Item = Notifier>) {
+    async fn attach_notifiers(&self, new_notifiers: Vec<Notifier>) {
         let mut queue = self.inner.queue.write().await;
+        let new_checkpoint = new_notifiers
+            .iter()
+            .any(|notifier| notifier.bound_to_checkpoint_barrier());
         match queue.front_mut() {
-            Some((_, notifiers)) => notifiers.extend(new_notifiers),
+            Some(Scheduled {
+                notifiers,
+                checkpoint,
+                ..
+            }) => {
+                notifiers.extend(new_notifiers);
+                *checkpoint = *checkpoint || new_checkpoint;
+            }
             None => {
                 // If no command scheduled, create periodic checkpoint barrier by default.
-                queue.push_back((Command::checkpoint(), new_notifiers.into_iter().collect()));
+                queue.push_back(Scheduled {
+                    notifiers: new_notifiers.into_iter().collect(),
+                    command: Command::checkpoint(),
```

Comment:
I named the periodic one with `checkpoint` as I was to cover the checkpoint flag inside the command. If we decide to put it in `Scheduled`, I think we'd better give a name other than `checkpoint` to this function.